File: tests/test_voice_stream.py
import asyncio
import websockets
import json
import wave
import sys
import os
import numpy as np

# Add project root to path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from config_system import config


async def test_voice_streaming():
    """
    Simulates a browser client streaming audio to the backend.
    """
    uri = f"ws://127.0.0.1:{config.voice_port}"
    print(f"Connecting to {uri}...")
    try:
        async with websockets.connect(uri) as websocket:
            # 1. Wait for Ready
            msg = await websocket.recv()
            print(f"< {msg}")
            # 2. Generate Synthetic Audio (Sine Wave "Beep")
            # We use a simple 440Hz tone. Whisper won't transcribe it as English text usually,
            # but it validates the pipeline processes audio without crashing.
            # Ideally we'd use a real wav file, but generating one is self-contained.
            print("> Streaming synthetic audio...")
            sample_rate = 16000
            duration = 3  # seconds
            frequency = 440
            t = np.linspace(0, duration, int(sample_rate * duration), False)
            note = np.sin(frequency * t * 2 * np.pi)
            audio = (note * 32767).astype(np.int16).tobytes()

            # Chunk it (0.25s chunks)
            chunk_size = int(sample_rate * 0.25 * 2)  # 2 bytes per sample

            for i in range(0, len(audio), chunk_size):
                chunk = audio[i : i + chunk_size]
                await websocket.send(chunk)
                await asyncio.sleep(0.1)  # Simulate real-time

                # Check for partials (non-blocking recv is tricky in simple loop, so we just send)

            print("> Audio finished. Sending Stop command...")
            await websocket.send(json.dumps({"command": "stop"}))

            # 3. Read Final Result
            while True:
                response = await websocket.recv()
                data = json.loads(response)
                print(f"< {data}")
                if data.get("is_final") or data.get("status") == "cleared":
                    break

            print("✅ Test Passed: Stream completed.")

    except Exception:
        print(f"❌ Test Failed: {e}")
        sys.exit(1)


if __name__ == "__main__":
    try:
        asyncio.run(test_voice_streaming())
    except KeyboardInterrupt:
        pass

File: rust_output/src/test_voice_stream.rs
use anyhow::{Result, Context};
use crate::config_system::{config};
use std::collections::HashMap;
use tokio;

/// Simulates a browser client streaming audio to the backend.
pub async fn test_voice_streaming() -> Result<()> {
    // Simulates a browser client streaming audio to the backend.
    let mut uri = format!("ws://127.0.0.1:{}", config::voice_port);
    println!("Connecting to {}...", uri);
    // try:
    {
        let mut websocket = websockets.connect(uri);
        {
            let mut msg = websocket.recv().await;
            println!("< {}", msg);
            println!("{}", "> Streaming synthetic audio...".to_string());
            let mut sample_rate = 16000;
            let mut duration = 3;
            let mut frequency = 440;
            let mut t = numpy.linspace(0, duration, (sample_rate * duration).to_string().parse::<i64>().unwrap_or(0), false);
            let mut note = numpy.sin((((frequency * t) * 2) * numpy.pi));
            let mut audio = (note * 32767).astype(numpy.int16).tobytes();
            let mut chunk_size = ((sample_rate * 0.25_f64) * 2).to_string().parse::<i64>().unwrap_or(0);
            for i in (0..audio.len()).step_by(chunk_size as usize).iter() {
                let mut chunk = audio[i..(i + chunk_size)];
                websocket.send(chunk).await;
                asyncio.sleep(0.1_f64).await;
            }
            println!("{}", "> Audio finished. Sending Stop command...".to_string());
            websocket.send(serde_json::to_string(&HashMap::from([("command".to_string(), "stop".to_string())])).unwrap()).await;
            while true {
                let mut response = websocket.recv().await;
                let mut data = serde_json::from_str(&response).unwrap();
                println!("< {}", data);
                if (data.get(&"is_final".to_string()).cloned() || data.get(&"status".to_string()).cloned() == "cleared".to_string()) {
                    break;
                }
            }
            println!("{}", "✅ Test Passed: Stream completed.".to_string());
        }
    }
    // except Exception as _e:
}
